### PR TITLE
Fix OOM when reading files that don't fit into memory

### PR DESCRIPTION
**Pull request by RobertLucian**
When running `cortex deploy -e aws`, if the project's files don't fit into memory, then an OOM error is returned and it panics. In Go, when that happens, the panic is non-recoverable. Because of this, the program must ensure it doesn't encounter an OOM.

Wanted to use [runtime.ReadMemStats](https://golang.org/pkg/runtime/#ReadMemStats) to get the available allocable memory (looking at _HeapIdle_), but after running a few tests, it doesn't look like it's giving what we need. The stats are mainly about what the Go program uses (with GC and all that) and not about the system's available memory.

And it wasn't like we could just read the stats from `/proc` because that's platform-specific. [shirou/gopsutil](https://github.com/shirou/gopsutil) appears to be a well-maintained project that's cross-platform and gives us what we need.

---

checklist:

- [x] run `make test` and `make lint`
- [x] test manually (i.e. build/push all images, restart operator, and re-deploy APIs)

### CODE REVIEW

File: pkg/lib/files/files.go [hunk 1]
```diff
@@ -106,6 +111,19 @@ func ReadFileBytesErrPath(path string, errMsgPath string) ([]byte, error) {
 		return nil, err
 	}
 
+	var fileSizeBytes int64
+	if fileSizeBytes, err = CheckFileSizeBytes(path); err != nil {
```
[the remark this answers is not in the document]
Comment (RobertLucian) [reply pkg/lib/files/files.go]: Good call!

File: pkg/lib/files/files.go [hunk 2]
```diff
@@ -301,6 +319,23 @@ func CheckFileErrPath(path string, errMsgPath string) error {
 	return nil
 }
 
+func CheckFileSizeBytes(path string) (int64, error) {
```
[the remark this answers is not in the document]
Comment (RobertLucian) [reply pkg/lib/files/files.go]: Yep! Done!

File: pkg/lib/files/files.go [hunk 3]
```diff
@@ -527,6 +532,22 @@ func IgnoreSpecificFiles(absPaths ...string) IgnoreFn {
 	}
 }
 
+func ErrorOnBigFiles(path string, fi os.FileInfo) (bool, error) {
```
Comment (vishalbollu) [inline pkg/lib/files/files.go:535]: I would restructure this function to follow the signature for `GitIgnoreFn` to return (IgnoreFn, error) and it takes in maxFileSize as a parameter.

Comment (RobertLucian) [reply pkg/lib/files/files.go:535]: I think we should only return `IgnoreFn` and not `(IgnoreFn, error)`. There shouldn't be any error returned from a function like `ErrorOnBigFilesFn`. As for its parameters, I think we actually need 2 of them: `maxFileSizeBytes` and `maxMemoryUsagePercent`. This way, we also get rid of 2 non-`files.go`-specific variables.

File: pkg/lib/files/errors.go [hunk 4]
```diff
@@ -87,6 +90,35 @@ func ErrorFileAlreadyExists(path string) error {
 	})
 }
 
+func ErrorInsufficientMemoryToReadFile(path string, fileSizeBytes, availableMemBytes int64) error {
```
Comment (vishalbollu) [inline pkg/lib/files/errors.go:93]: This function doesn't accept path

File: pkg/lib/files/files.go [hunk 5]
```diff
@@ -552,13 +555,44 @@ func PromptForFilesAboveSize(size int, promptMsgTemplate string) IgnoreFn {
...
 		return false, nil
 	}
 }
 
+func ErrorOnBigFilesFn(maxFileSizeBytes int64, maxMemoryUsagePercent float64) IgnoreFn {
+	return func(path string, fi os.FileInfo) (bool, error) {
+		if !fi.IsDir() {
+			fileSizeBytes := fi.Size()
+			virtual, _ := mem.VirtualMemory()
+			if float64(fileSizeBytes) > float64(virtual.Available) ||
```
Comment (vishalbollu) [inline pkg/lib/files/files.go:570]: `float64(fileSizeBytes) > float64(virtual.Available) ||` condition may not be necessary

**Pull request merged**